### PR TITLE
add block_endorsement_repository

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+   * @invariant WriteBatch is always in valid state.
+   * @invariant WriteBatch does not modify on-disk storage after put/remove
+   * operations. It does, when \p BlockEndorsmentRepository::commit is executed
+   * on this batch.
+   */
+  struct WriteBatch {
+    virtual ~WriteBatch() = default;
+
+    /**
+     * Write a single endorsement which endorses a endorsed_block and contains
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: an endorsed

Comment (blackyblack) [reply include/veriblock/storage/block_endorsment_repository.hpp]: and is contained

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+     * in the containig_block. If endorsement exist, db overwrites it. it.
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: in the containing_block

Comment (blackyblack) [reply include/veriblock/storage/block_endorsment_repository.hpp]: Remove extra 'it.'

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+     * @return true if we overwrite existing data, false otherwise.
+     */
+    virtual bool put(const stored_endorsement_t& endorsement,
+                     const stored_block_t& containing_block,
+                     const stored_block_t& endorsed_block) = 0;
+
+    /**
+     * Remove a single endorsement with all references to the containing_blocks
+     * and endorsed_blocks from storage identified by endorsement hash.
+     * @param hash endoresement hash
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: endorsement

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+     * @return true if removed, false if no such element found.
+     */
+    virtual bool removeEndorsmentByHash(const endorsement_hash_t& hash) = 0;
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: removeEndorsement

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    /**
+     * Remove a single containing_block with all references to the endorsement
+     * and romove all endorsements and endorsed_blocks from storage identified
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: remove

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    virtual bool removeEndorsedBlockByHash(const block_hash_t& hash) = 0;
+
+    /**
+     * Clear batch from any modifying operations.
+     */
+    virtual void clear() = 0;
+  };
+
+  /**
+   * Write a single endorsement which endorses a endorsed_block and contains
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: an endorsed_block

Comment (blackyblack) [reply include/veriblock/storage/block_endorsment_repository.hpp]: and is contained in the containing_block

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+   * in the containig_block. If endorsement exist, db overwrites it. it.
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: Remove extra 'it'

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+   * @param endorsement to be written in a batch
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: This is not a batch method

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+   * @param containing_block to be written in a batch
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: Same

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+   * @param endorsed_block to be written in a batch
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: And same

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                   const stored_block_t& containing_block,
+                   const stored_block_t& endorsed_block) = 0;
+
+  /**
+   * Remove a single endorsement with all references to the containing_blocks
+   * and endorsed_blocks from storage identified by endorsement hash.
+   * @param hash endoresement hash
+   * @return true if removed, false if no such element found.
+   */
+  virtual bool removeEndorsmentByHash(const endorsement_hash_t& hash) = 0;
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: removeEndorsementByHash

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+  /**
+   * Remove a single containing_block with all references to the endorsement
+   * and romove all endorsements and endorsed_blocks from storage identified
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: remove

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+   * @param hash block hash
+   * @return true if removed, false if no such element found.
+   */
+  virtual bool removeEndorsedBlockByHash(const block_hash_t& hash) = 0;
+
+  /**
+   * Retrieve endorsements from the given containing blocks that endorse the
+   * given endorsed block
+   * @param endorsed_block_hash[in] the endorsed block hash
+   * @param contaning_block_hashes[in] the containing block hashes
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: containing_block_hashes

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+   * @param endorsements[out] if non-null, endorsements is written here. If null
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: are written

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+   * is passed, out arguments is ignored.
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: out argument

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+   * @return number of endorsements appended to output vector.
+   */
+  virtual size_t getEndorsements(
+      const block_hash_t& endorsed_block_hash,
+      const std::vector<block_hash_t>& contaning_block_hashes,
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: containing_block_hashes

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+      std::vector<stored_endorsement_t>* endorsements) = 0;
+
+  /**
+   * Retrieve Endorsements from the given containing block
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: endorsements

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+   * @param containing_block_hash[in] the containing block hash
+   * @param endorsements[out] if non-null, endorsements is written here. If null
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: are written

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+   * is passed, out arguments is ignored.
```
Comment (blackyblack) [inline include/veriblock/storage/block_endorsment_repository.hpp]: out argument

File: include/veriblock/storage/block_endorsment_repository.hpp [hunk 20]
```diff
@@ -0,0 +1,152 @@
+#ifndef ALT_INTEGRATION_VERIBLOCK_BLOCK_STORAGE_ENDORSMENT_REPOSITORY_HPP
+#define ALT_INTEGRATION_VERIBLOCK_BLOCK_STORAGE_ENDORSMENT_REPOSITORY_HPP
+
+#include <memory>
+#include <vector>
+
+namespace VeriBlock {
+
+template <typename Endorsement, typename AltChainBlock>
```
Comment (Warchant) [inline include/veriblock/storage/block_endorsment_repository.hpp]: AltChainBlock should be probably just a struct like this:
```
struct AltChainBlock {
  std::vector<uint8_t> hash;
  uint32_t timestamp;
  uint32_t height; 
}
```

**Comment by Warchant** [comment]
Will reopen it later

**Pull request closed without merging**